### ISSUE TITLE
Replicator stopped  and giving 409 during the system test execution for doc update

### ISSUE THREAD

**Issue by hrajput89** [couchbase/couchbase-lite-net#993]
After replicating 1M docs to SG/CBS, when the system test tried to update the docs the test server threw 409 and later while adding docs to CBL replicator responded with Stopped status.

CBL client logs - 

[cbl_client.log](https://github.com/couchbase/couchbase-lite-net/files/1808104/cbl_client.log)

The cluster is set in local network. Although I'm re-running the test just to be sure. 

Here is the test command -
pytest -rsx -s --mode=cc --server-version=5.5.0-1992 --sync-gateway-version=2.0.0-823 --liteserv-platform=dotnet --liteserv-version=2.0.0-0b636 --liteserv-host=192.168.0.145 --liteserv-port=55555 --xattrs --create-db-per-suite="suite_db" --create-db-per-test="test_db"  --timeout=216000 --skip-provisioning -k test_system testsuites/CBLTester/System_Test/test_system_testing.py

**Comment by borrrden** [comment]
Can you get a reproduction of this that doesn't take 1h+ to run?

**Comment by borrrden** [comment]
Can you be more specific about what is happening here?  A 409 is a valid thing to happen during replication.  I don't see anything in the logs I received via email except that there were some 409 (attempted conflict to Sync Gateway) and then a timeout happened.  Can this be reproduced in a simpler way?

Furthermore I don't see anywhere in the logs on this ticket that indicate that "the test server threw 409" so I'm having a hard time figuring out what the expected behavior is here other than "the replication should not stop" and the reason it stopped is unclear.

**Comment by hrajput89** [comment]
Jim, I can't reproduce this scenarios in less than an hour as I see issue with system testing of 1M docs. I tried that same system test for 500k docs and we didn't observe any issue. Just to give you a brief about what system test does, below is the description - 

```
Below is the description of system testing test - 
1.  Add 1M docs in bulk (10000 docs per 100 iteration) sequentially to CBL app. Doc id is of format “cbl_<db_name>_<num>”  - “cbl_test_db_0”
2. Start replication for CBP app to connect to SG/CBS with PUSH and PULL replication. Wait for the replication to finish and then check the doc count in db is same as num of docs added
3. Iterate over all the operations below for the given period of time
4. Update some random docs on SG and let the replication finish on devices. Verify that the doc count matches with the length of doc_id list.
5. Update docs in bulk on CBL App and check for the replication status. Verify that the doc count matches with the length of doc_id list.
6. Delete some random docs on SG. Remove the doc ids for the deleted docs for the doc_ids list
7. Delete docs in bulk on CBL app and check the replication status. Verify that the doc count matches with the length of doc_id list.  
8. Add few docs to CBL app and check the replication status. Verify that the doc count matches with the length of doc_id list.
9. Repeat until timeout to step 3. 
10. After timeout close all the replication and finish test.
```

The replication gets stopped at step 8 CBL client run snippet below (cbl_client.log attached file). From there onwards replication never recovered.

```
deleting 888 docs on CBL
Got response:
Got response: "busy"
Activity level: busy
Got response:
Got response: "busy"
Activity level: busy
Got response:
Got response: "busy"
Activity level: busy
Got response:
Got response: "busy"
Activity level: busy
Got response:
Got response: "busy"
Activity level: busy
Got response:
Got response: "idle"
Got response: L200115871
Got response: L200115871
GET http://192.168.0.149:4985/db/_all_docs?include_docs=true 200
creating 1000 docs on CBL
Got response:
Got response: "stopped"
Activity level: stopped
```

**Comment by borrrden** [comment]
Can you set up a test that will start with CBL and Sg full so that this case can be run without waiting for 1 hour?  Nothing in what you pasted says anything about a reason (it just says “stopped”) and if the logs are so huge that to can’t paste them here then just take the last 1000 lines or so with the info log level at least.

**Comment by hrajput89** [comment]
@borrrden, I started the test with SG and CBL full and did failed again for the same reason. 
PFA logs (last 1000 lines from Testserver logs)
[testserver.log](https://github.com/couchbase/couchbase-lite-net/files/1812182/testserver.log)

**Comment by borrrden** [comment]
Is this the same cluster that is being used for the Android system test?  The logs say that the stop was caused by a timeout from Sync Gateway.

**Comment by hrajput89** [comment]
The cluster for this run is not same, but the issue is reproducible on that cluster too (bar the testserver machine).

**Comment by borrrden** [comment]
I missed the part where you said you started the test with them full.  Can I have that test setup to try?

**Comment by borrrden** [comment]
This looks like an actual timeout from what I can tell.  I ran the test twice, and both times the test was "stuck" for 5 minutes before this happened.  During the time Sync gateway was going crazy writing hundreds of thousands of lines of this:

> 2018-03-21 04:40:41.628234 I | http: multiple response.WriteHeader calls
2018-03-21T04:40:41.629Z CRUD+: Couldn't serialize document body, HTTP client closed connection

This appears to be in response to the python client's \_all\_docs call.  While the client is waiting for this to happen, the five minute timeout period happens and the replication stops.  The replication should not stop, but that is a LiteCore issue and not the focus of this test...

Even after the replicator stops (most of the time) the python client is still stuck waiting for something while Sync Gateway runs at 100% CPU spitting out those lines.  I cannot rely on this test in this condition.

**Comment by borrrden** [comment]
I set this up in a proper environment (running on AWS with 16GB of RAM and 4 vCPU with both SG and CBS installed) and not only did the replication complete in less than 20 minutes instead of more than an hour, but I did not observe the issue.  I do notice something that backs up my theory, however.  Even in this environment the call to `_all_docs` takes nearly 3 minutes to complete since it specifies to include the bodies of 1M+ docs in the result:

> 2018-03-22T04:29:44.450Z HTTP:  #18978: GET /db/_all_docs?include_docs=true (as ADMIN)
2018-03-22T04:30:39.293Z Cache+: Pruned 101 old entries from channel "*"
2018-03-22T04:30:39.293Z Cache+: Pruned 101 old entries from channel "ABC"
2018-03-22T04:32:42.306Z HTTP+: #18978:     --> 200   (177856.2 ms)

This could easily pass the 5 minute mark if running on starved hardware, during which time the replication is just waiting with no information.  I ran it once on my laptop before I had to quit it because it made my system unresponsive running with so many VMs.  I noticed that it timed out like before, but then reconnected as expected.  I think that this is a flaw in the test (requesting all 1M docs from SG after every operation WITH their bodies, and then pretty much ignoring them).  However, be that as it may, the replication should continue now.
